### ISSUE TITLE
Does IB Framework really work?

### ISSUE THREAD

**Issue by kuaile258890** [qiuzx2/CIBHash#6]
Hello, this work introduces information theory into contrastive learning, and I believe it is an enlightening piece of work. I think the introduction of information theory is the most brilliant part of this paper. However, the ablation experiments in the paper show that the IB Framework has almost no improvement in performance (only about 0.01 MAP increase). In my own experiments, CLHash (with w set to 0) achieved a 0.6459 result on the CIFAR10 dataset with 64 bits, while CIBHash (with w set according to the paper at 0.001) resulted in 0.6454. There is virtually no difference between the two. Additionally, your setting of w to 0.001 is not a very large number. Can IB really help the model learn better?

**Comment by zexuanqiu** [comment]
Hi, in our previous experiments, indeed, the KL loss part in IB has limited improvement. This is also one of the limitations of our paper. But I think there is actually some room for improvement in the performance improvement of IB framework. For example, KL divergence may not be a tight mutual information estimator, and some other mutual information estimation methods (e.g., CLUB, ICML2020) may possibly achieve better results.

**Comment by kuaile258890** [comment]
> Hi, in our previous experiments, indeed, the KL loss part in IB has limited improvement. This is also one of the limitations of our paper. But I think there is actually some room for improvement in the performance improvement of IB framework. For example, KL divergence may not be a tight mutual information estimator, and some other mutual information estimation methods (e.g., CLUB, ICML2020) may possibly achieve better results.

Thank you for your response. Although the improvement in KL loss is limited, the introduction of mutual information still provides a new perspective for contrastive learning. Therefore, I believe this is still an insightful work. At the same time, I would like to remind you that there might be an error in the implementation of the forward function in your CIBHash.py. That is,
`z_i = hash_layer(prob_i - 0.5)`
should be changed to
`z_i = hash_layer(prob_i - torch.empty_like(prob_i).uniform_().to(prob_i.device))`
The same applies to z_j. Because you mentioned in the paper that 0.5 is only used during the testing phase, while during the training phase, the code should use u which is sampled from a uniform distribution [0,1].
This part of the code has been modified in the subsequent work CGHash from your lab.

**Comment by zexuanqiu** [comment]
Hi, thank you for your modification! Your modification is correct. I'll correct it later.

**Comment by FuugaMo** [comment]
Hello, I have been trying to reproduce the results of this paper recently. I am curious if you made any additional modifications besides the two adjustments in the `forward` function? Since that I am encountering issues with gradient explosion. @kuaile258890 It would help a lot!

**Comment by kuaile258890** [comment]
> Hello, I have been trying to reproduce the results of this paper recently. I am curious if you made any additional modifications besides the two adjustments in the `forward` function? Since that I am encountering issues with gradient explosion. @kuaile258890 It would help a lot!

No, I haven't made any changes. In fact, apart from the IB issue I mentioned, this is a very replicable piece of work, and the results can be replicated by simply inputting commands. I noticed you mentioned your CIFAR-10 dataset coming from the CIFAR-10 official website, while the dataset I used for replication was downloaded automatically by the datasets.

**Comment by FuugaMo** [comment]
Got it, though I have tried to use the dataset downloaded automatically by the code, it still does not work. May I ask about your GPU and cuda version? Or can I have your requirements.txt to further check the environment difference? @kuaile258890

**Comment by kuaile258890** [comment]
> Got it, though I have tried to use the dataset downloaded automatically by the code, it still does not work. May I ask about your GPU and cuda version? Or can I have your requirements.txt to further check the environment difference? @kuaile258890

My machine is equipped with 8 NVIDIA GeForce RTX 2080 Ti GPUs, but only one GPU is actually utilized. The CUDA version is 10.2, and the Driver Version is 440.33.01. Below are the versions of various libraries in my environment, hoping it will be helpful to you.

torch                        1.12.1+cu102
torchvision                  0.13.1+cu102
Pillow                       9.5.0
opencv-python                4.5.1.48

**Comment by FuugaMo** [comment]
Thank you for the information! It turns out that this problem is caused by **lack of GPU memory and some kind of unknown bug in some previous CUDA versions?**

I have tried:
- Setting Learning Rate to 0: Did not solve the issue.
- Switching to CPU Execution: Loss explosion issue did not occur on CPU, but it was tooooo slow.
- Modifying forward Function Code: Followed your code modification, but problem persisted.
- Upgrading PyTorch and Related Libraries: Tried upgrading to versions with 'cuda' suffix (from Torch ==1.4.0 to Torch==1.5.0+cuda102), but loss explosion occurred earlier (before eval, in the first epoch).
- Downgrading CUDA Version: Attempted downgrade to CUDA 9.2, encountered new errors. Since my GPU 3070-8G's compute capability  is 8.5 but CUDA 9.2 only support it in the range 3-7.
- Changing Data Type Precision to float64: Consumed excessive memory, necessitating reduction of batch size to 16.
Decreasing Batch Size: Set batch size to 2, mitigated the problem, but it hardly improve the model after evaluating. A batch size of 4 leads to the loss NaN problem.
- Migrating to Alternative Environment: Shifted to a Colab environment, indicating a possible issue with GPU memory constraints.
- Upgrading CUDA Version to 11.1: Upgraded to a newer CUDA version, encountered MemoryError.

Here's some information of difference performance on different GPUs that I have gathered or tried:

My Machine:
- GPU: NVIDIA GeForce RTX 3070 8G
- CUDA Version: 
    - Known NaN issues occurred with CUDA 10.1; 
    - Compute capability compatibility problem in CUDA 9.2;
    - MemoryError problem in CUDA 11.1; 

Your Machine:
- GPU: 8 NVIDIA GeForce RTX 2080 Ti (11GB perhaps?)
- CUDA Version: 10.2
- Driver Version: 440.33.01
- Success, no NaN problem.

Colab Environment:
- GPU: NVIDIA T4 (Tesla T4)
- VRAM: 15GB
- CUDA Version: 12+
- The graph in Colab shows that the model occupied 9.2GB of all GPU RAM.
- Success, no NaN problem.

A40 GPU Environment:
- GPU: NVIDIA A40
- VRAM: Not specified
- CUDA Version: Not specified, but known to work fine in this environment.
- Success, no NaN problem.

In conclusion, I suspect that the MemoryError issue was not properly handled in older versions of CUDA, as reported in https://github.com/ultralytics/ultralytics/issues/5294: CUDA errors like "out of memory" may also lead to NaN results.

My suspicion is that older versions of CUDA may lack a proper error handling mechanism for NaN caused by insufficient memory, however, I have too few evidence of this problem. If anyone knows more specific detail of this problem, please contact me!
